File: src/codes/hardcoded-variability/py/rfc_bridge.py
from abc import ABC, abstractmethod

# 実装側（Bridge の下部）
class Notifier(ABC):
    @abstractmethod
    def send_message(self, to: str, content: str):
        pass

class EmailNotifier(Notifier):
    def send_message(self, to: str, content: str):
        print("件名: お知らせ")
        print(f"本文: {content}")
        print(f"宛先: {to}@example.com")

class SlackNotifier(Notifier):
    def send_message(self, to: str, content: str):
        print(f"Slack宛: {to}")
        print(f"内容: {content}")

# 抽象側（Bridge の上部）
class Notification(ABC):
    def __init__(self, notifier: Notifier):
        self.notifier = notifier

    @abstractmethod
    def notify(self, user: str, message: str):
        pass

class NormalNotification(Notification):
    def notify(self, user: str, message: str):
        self.notifier.send_message(user, message)

class UrgentNotification(Notification):
    def notify(self, user: str, message: str):
        self.notifier.send_message(user, f"[緊急] {message}")

# 利用例
email_urgent = UrgentNotification(EmailNotifier())
email_urgent.notify("hiroshi", "サーバーが落ちました")

slack_normal = NormalNotification(SlackNotifier())
slack_normal.notify("hiroshi", "定例ミーティングは15時です")

File: src/codes/hardcoded-variability/ts/rfc_bridge.ts
// 実装クラス（通知手段）
interface Notifier {
  sendMessage(to: string, content: string): void;
}

class EmailNotifier implements Notifier {
  sendMessage(to: string, content: string): void {
    console.log("件名: お知らせ");
    console.log(`本文: ${content}`);
    console.log(`宛先: ${to}@example.com`);
  }
}

class SlackNotifier implements Notifier {
  sendMessage(to: string, content: string): void {
    console.log(`Slack宛: ${to}`);
    console.log(`内容: ${content}`);
  }
}

// 抽象クラス（通知処理）
abstract class Notification {
  constructor(protected notifier: Notifier) {}

  abstract notify(user: string, message: string): void;
}

// 拡張された抽象クラス
class NormalNotification extends Notification {
  notify(user: string, message: string): void {
    this.notifier.sendMessage(user, message);
  }
}

class UrgentNotification extends Notification {
  notify(user: string, message: string): void {
    this.notifier.sendMessage(user, `[緊急] ${message}`);
  }
}

// 利用例
const emailUrgent = new UrgentNotification(new EmailNotifier());
emailUrgent.notify("hiroshi", "サーバーが落ちました");

const slackNormal = new NormalNotification(new SlackNotifier());
slackNormal.notify("hiroshi", "定例ミーティングは15時です");
